File: frontend/src/components/today/ritualCueSanitizer.ts
/**
 * Убираем из UI ритуала «псевдотексты»: сырые slug'и, метки-рубрики вместо действий,
 * шаблон «Не заходить в линию 'general'» из кэша/старого бэка.
 */
import { isDiscardableMorningFocus } from "@/components/today/todayRitualCopy";

/** Паритет `ritual_cue_sanitize._TOPIC_LABELS_NOT_ACTIONS` (O3). */
export const RU_TOPIC_LABELS_NOT_ACTIONS = new Set([
  "смысл и коммуникация",
  "смысл и коммуникации",
  "смысл и коммуникацию",
  "общий фокус дня",
  "общий фон дня",
  "общение и контакт",
  "смысл дня",
  "контекст дня",
  "рамка дня",
  "общая картина",
  "картина дня",
  "настрой на день",
  "тональность дня",
  "вектор дня",
  "форма дня",
  "сигнал дня",
  "паттерн дня",
]);

const TOPIC_LABELS_NOT_ACTIONS = RU_TOPIC_LABELS_NOT_ACTIONS;

/** O3: заголовок day_layer не должен быть только рубрикой (паритет бэкенда `is_ru_abstract_topic_headline`). */
export function isRuAbstractTopicHeadline(text: string | null | undefined): boolean {
  const t = String(text ?? "")
    .trim()
    .toLowerCase();
  if (!t) return true;
  return TOPIC_LABELS_NOT_ACTIONS.has(t);
}

const SLUG_TOPIC_JUNK = new Set([
  "general",
  "overall",
  "dialogue",
  "communication",
  "mixed",
  "none",
  "other",
  "default",
]);

const SLUG_TO_RU: Record<string, string> = {
  general: "общий фон дня",
  love: "любовь и близость",
  relations: "отношения",
  career: "работа и карьера",
  work: "работа и карьера",
  money: "деньги и границы",
  family: "семья и дом",
  home: "семья и дом",
  body: "тело и восстановление",
  health: "тело и восстановление",
  dialogue: "общение и контакт",
  communication: "общение и контакт",
  decision: "решение, которое надо принять",
  identity: "линия про себя",
  self: "линия про себя",
};

/** Паритет `day_narrative_brief_v0` / `ritual_cue_sanitize._HEAD_TOPIC_SLUG_RU` (O5). */
const HEAD_TOPIC_SLUG_RU: Record<string, string> = {
  general: "общий фон дня",
  body: "тело и энергия",
  money: "деньги",
  dialogue: "общение и контакт",
  family: "семья и дом",
  career: "работа и дела",
  love: "близость и отношения",
};

/** Паритет `day_narrative_brief_v0._MOOD_ID_RU` / `ritual_cue_sanitize._MOOD_SLUG_TO_RU` (O5). */
const MOOD_SLUG_RU: Record<string, string> = {
  calm: "спокойно",
  anxious: "тревожно",
  tired: "устало",
  driven: "в драйве",
  irritated: "раздражённо",
  other: "другое",
  motivated: "в драйве",
  confused: "неясно",
  quiet_wish: "хочется тишины",
  move_wish: "хочется движения",
  heavy: "тяжело",
  hopeful: "с надеждой",
  distant: "на дистанции",
};

const QUOTED_SERVICE_SLUG_TO_RU: Record<string, string> = {
  ...SLUG_TO_RU,
  ...HEAD_TOPIC_SLUG_RU,
  ...MOOD_SLUG_RU,
};

const QUOTED_EN_SERVICE_SLUG_RE = /(?:['"]|«)([a-z][a-z0-9_]{0,31})(?:['"]|»)/gi;

/** O5: в RU-тексте не показываем сырые 'tired' / 'general' в кавычках. */
export function replaceQuotedEnSlugsForRuDisplay(text: string | null | undefined): string {
  const raw = String(text ?? "").trim();
  if (!raw) return "";
  return raw.replace(QUOTED_EN_SERVICE_SLUG_RE, (full, g1: string) => {
    const slug = String(g1 || "")
      .trim()
      .toLowerCase();
    const label = QUOTED_SERVICE_SLUG_TO_RU[slug];
    return label ? `«${label}»` : full;
  });
}

const DO_NOT_ENTER_CHAOS_RU =
  "Осторожнее, если день начинает скатываться в хаос, резкие реакции и потерю своего ритма — держи одну линию, не хватайся за всё сразу.";

export function humanizeFocusSlugForUi(slug: string): string {
  const k = slug.trim().toLowerCase();
  if (SLUG_TO_RU[k]) return SLUG_TO_RU[k];
  if (/^[a-z][a-z0-9_]{0,22}$/.test(k)) return "узкая тема дня";
  return slug.trim();
}

/** Строка «что делать» не должна быть только рубрикой/темой без прикладного действия. */
export function isGarbageRitualActionCue(line: string | null | undefined): boolean {
  const raw = (line || "").trim();
  if (!raw) return true;
  if (isDiscardableMorningFocus(raw)) return true;
  const t = raw.toLowerCase();
  if (TOPIC_LABELS_NOT_ACTIONS.has(t)) return true;
  if (/^[\s_a-z]+$/.test(t) && t.length <= 32) return true;
  return false;
}

/**
 * Починка кэшированных/старых do_not_enter с кавычками и англ. slug.
 * Если паттерн узнаётся — отдаём одну нормальную фразу без «линии 'general'».
 */
export function repairRitualDoNotEnterLine(raw: string | null | undefined): string {
  const t = replaceQuotedEnSlugsForRuDisplay(raw);
  if (!t) return "";
  const quoted = t.match(/['«]([a-z][a-z0-9_]{0,24})['»]/i);
  if (quoted) {
    const slug = (quoted[1] || "").trim().toLowerCase();
    if (isDiscardableMorningFocus(slug) || SLUG_TOPIC_JUNK.has(slug)) {
      return DO_NOT_ENTER_CHAOS_RU;
    }
    const label = humanizeFocusSlugForUi(quoted[1]);
    return `Осторожнее с темой «${label}», если она начинает проживаться как хаос, резкие реакции и потеря своего ритма.`;
  }
  if (/\bgeneral\b/i.test(t) && /линию|линия/i.test(t)) {
    return DO_NOT_ENTER_CHAOS_RU;
  }
  return t;
}

/** Паритет `ritual_cue_sanitize._LLM_META_NEEDLES` / `strip_llm_meta_commentary` (Python). */
export const LLM_META_NEEDLES: readonly string[] = [
  "не дублирую",
  "не дублируем",
  "я не дублиру",
  "чтобы экран не перегруж",
  "экран не перегружа",
  "карта и число остаются",
  "не дублирую их",
  "не дублируем их",
  "в сводке и в",
  "чтобы не перегруж",
  "чтобы не дублировать",
  "не дублирую информацию",
  "не дублируем информацию",
  "не повторяю блок",
  "не повторяю уже сказанное",
  "как просили в промпте",
  "как указано в задании",
  "в рамках формата ответа",
  "по требованиям к ответу",
  "согласно инструкции для модели",
  "убираю дублирование",
  "исключил дублирование",
  "дублирование с предыдущим",
  "уже было в предыдущем блоке",
  "из предыдущего абзаца",
  "as per the prompt",
  "as instructed, i will not",
  "i won't repeat the",
  "to avoid duplication",
  "avoiding repeating",
  "not repeating the card",
  "not repeating the number",
];

/** Убираем предложения с мета-комментарием модели (не пользовательский смысл дня). */
export function stripLlmMetaCommentary(text: string | null | undefined): string {
  const raw = String(text ?? "").trim();
  if (!raw) return "";
  const low = raw.toLowerCase();
  if (!LLM_META_NEEDLES.some((n) => low.includes(n))) return raw;
  const parts = raw.split(/(?<=[.!?])\s+/);
  const kept = parts
    .map((p) => p.trim())
    .filter((pl) => pl.length > 0 && !LLM_META_NEEDLES.some((n) => pl.toLowerCase().includes(n)));
  return kept.join(" ").trim();
}

File: android/TodayFlow/app/src/main/java/com/todayflow/today/RitualCueSanitizer.kt
package com.todayflow.today

/**
 * Паритет с `frontend/src/components/today/ritualCueSanitizer.ts` и iOS `TodayRitualCueSanitizer.swift`.
 * Использовать, когда на Android появится UI ритуала «Сегодня».
 */
object RitualCueSanitizer {
    private const val CHAOS_LINE =
        "Осторожнее, если день начинает скатываться в хаос, резкие реакции и потерю своего ритма — держи одну линию, не хватайся за всё сразу."

    private val slugToRu: Map<String, String> =
        mapOf(
            "general" to "общий фон дня",
            "love" to "любовь и близость",
            "relations" to "отношения",
            "career" to "работа и карьера",
            "work" to "работа и карьера",
            "money" to "деньги и границы",
            "family" to "семья и дом",
            "home" to "семья и дом",
            "body" to "тело и восстановление",
            "health" to "тело и восстановление",
            "dialogue" to "общение и контакт",
            "communication" to "общение и контакт",
            "decision" to "решение, которое надо принять",
            "identity" to "линия про себя",
            "self" to "линия про себя",
        )

    private val headTopicSlugRu: Map<String, String> =
        mapOf(
            "general" to "общий фон дня",
            "body" to "тело и энергия",
            "money" to "деньги",
            "dialogue" to "общение и контакт",
            "family" to "семья и дом",
            "career" to "работа и дела",
            "love" to "близость и отношения",
        )

    private val moodSlugRu: Map<String, String> =
        mapOf(
            "calm" to "спокойно",
            "anxious" to "тревожно",
            "tired" to "устало",
            "driven" to "в драйве",
            "irritated" to "раздражённо",
            "other" to "другое",
            "motivated" to "в драйве",
            "confused" to "неясно",
            "quiet_wish" to "хочется тишины",
            "move_wish" to "хочется движения",
            "heavy" to "тяжело",
            "hopeful" to "с надеждой",
            "distant" to "на дистанции",
        )

    private val quotedServiceSlugToRu: Map<String, String> =
        buildMap {
            putAll(slugToRu)
            putAll(headTopicSlugRu)
            putAll(moodSlugRu)
        }

    private val quotedEnServiceSlug =
        Regex("""(?:['"]|«)([a-z][a-z0-9_]{0,31})(?:['"]|»)""", RegexOption.IGNORE_CASE)

    /** O5: паритет `ritual_cue_sanitize.replace_quoted_en_slugs_for_ru_display`. */
    @JvmStatic
    fun replaceQuotedEnSlugsForRuDisplay(raw: String?): String {
        val t = (raw ?: "").trim()
        if (t.isEmpty()) return ""
        return quotedEnServiceSlug.replace(t) { m ->
            val slug = m.groupValues[1].lowercase()
            val label = quotedServiceSlugToRu[slug]
            if (label != null) "«$label»" else m.value
        }
    }

    private val slugTopicJunk: Set<String> =
        setOf(
            "general",
            "overall",
            "dialogue",
            "communication",
            "mixed",
            "none",
            "other",
            "default",
        )

    /** Паритет `ritual_cue_sanitize._TOPIC_LABELS_NOT_ACTIONS` (O3). */
    private val topicOnly: Set<String> =
        setOf(
            "смысл и коммуникация",
            "смысл и коммуникации",
            "смысл и коммуникацию",
            "общий фокус дня",
            "общий фон дня",
            "общение и контакт",
            "смысл дня",
            "контекст дня",
            "рамка дня",
            "общая картина",
            "картина дня",
            "настрой на день",
            "тональность дня",
            "вектор дня",
            "форма дня",
            "сигнал дня",
            "паттерн дня",
        )

    /** O3: заголовок day_layer не только рубрика. */
    @JvmStatic
    fun isRuAbstractTopicHeadline(text: String?): Boolean {
        val t = (text ?: "").trim().lowercase()
        if (t.isEmpty()) return true
        return topicOnly.contains(t)
    }

    private val junkFocus: Set<String> =
        setOf(
            "general",
            "overall",
            "mixed",
            "none",
            "other",
            "default",
            "общее",
            "прочее",
            "другое",
            "без фокуса",
        )

    @JvmStatic
    fun humanizeFocusSlugForUi(slug: String): String {
        val k = slug.trim().lowercase()
        slugToRu[k]?.let { return it }
        if (Regex("^[a-z][a-z0-9_]{0,22}$").matches(k)) return "узкая тема дня"
        return slug.trim()
    }

    @JvmStatic
    fun isDiscardableMorningFocus(focus: String?): Boolean {
        val t = (focus ?: "").trim().lowercase()
        if (t.length < 2) return true
        if (junkFocus.contains(t)) return true
        if (Regex("^[a-z_]{1,20}$").matches(t)) return true
        return false
    }

    @JvmStatic
    fun isGarbageRitualActionCue(line: String?): Boolean {
        val raw = (line ?: "").trim()
        if (raw.isEmpty()) return true
        if (isDiscardableMorningFocus(raw)) return true
        val t = raw.lowercase()
        if (topicOnly.contains(t)) return true
        if (t.length <= 32 && Regex("^[\\s_a-z]+$").matches(t)) return true
        return false
    }

    private val quotedSlug = Regex("""['«]([a-z][a-z0-9_]{0,24})['»]""", RegexOption.IGNORE_CASE)

    @JvmStatic
    fun repairRitualDoNotEnterLine(raw: String?): String {
        val t = replaceQuotedEnSlugsForRuDisplay(raw)
        if (t.isEmpty()) return ""
        quotedSlug.find(t)?.groupValues?.getOrNull(1)?.let { slugRaw ->
            val slug = slugRaw.trim().lowercase()
            if (isDiscardableMorningFocus(slug) || slugTopicJunk.contains(slug)) {
                return CHAOS_LINE
            }
            val label = humanizeFocusSlugForUi(slugRaw)
            return "Осторожнее с темой «$label», если она начинает проживаться как хаос, резкие реакции и потеря своего ритма."
        }
        if (Regex("general", RegexOption.IGNORE_CASE).containsMatchIn(t) &&
            Regex("линию|линия", RegexOption.IGNORE_CASE).containsMatchIn(t)
        ) {
            return CHAOS_LINE
        }
        return t
    }

    /** Паритет `ritual_cue_sanitize.strip_llm_meta_commentary` / `stripLlmMetaCommentary` (TS). */
    private val llmMetaNeedles: List<String> =
        listOf(
            "не дублирую",
            "не дублируем",
            "я не дублиру",
            "чтобы экран не перегруж",
            "экран не перегружа",
            "карта и число остаются",
            "не дублирую их",
            "не дублируем их",
            "в сводке и в",
            "чтобы не перегруж",
            "чтобы не дублировать",
            "не дублирую информацию",
            "не дублируем информацию",
            "не повторяю блок",
            "не повторяю уже сказанное",
            "как просили в промпте",
            "как указано в задании",
            "в рамках формата ответа",
            "по требованиям к ответу",
            "согласно инструкции для модели",
            "убираю дублирование",
            "исключил дублирование",
            "дублирование с предыдущим",
            "уже было в предыдущем блоке",
            "из предыдущего абзаца",
            "as per the prompt",
            "as instructed, i will not",
            "i won't repeat the",
            "to avoid duplication",
            "avoiding repeating",
            "not repeating the card",
            "not repeating the number",
        )

    private val sentenceBoundarySplit = Regex("(?<=[.!?])\\s+")

    @JvmStatic
    fun stripLlmMetaCommentary(text: String?): String {
        val raw = (text ?: "").trim()
        if (raw.isEmpty()) return ""
        val low = raw.lowercase()
        if (llmMetaNeedles.none { low.contains(it) }) return raw
        return sentenceBoundarySplit
            .split(raw)
            .map { it.trim() }
            .filter { part ->
                part.isNotEmpty() && llmMetaNeedles.none { n -> part.lowercase().contains(n) }
            }
            .joinToString(" ")
            .trim()
    }
}
